File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/services/SuggestionServices.java
package com.expedia.bookings.services;

import java.util.List;

import com.expedia.bookings.data.SuggestionResultType;
import com.expedia.bookings.data.cars.Suggestion;
import com.expedia.bookings.data.cars.SuggestionResponse;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.squareup.okhttp.OkHttpClient;

import retrofit.RequestInterceptor;
import retrofit.RestAdapter;
import retrofit.client.OkClient;
import retrofit.converter.GsonConverter;
import rx.Observable;
import rx.Observer;
import rx.Scheduler;
import rx.Subscription;
import rx.functions.Func1;

public class SuggestionServices {

	private SuggestApi mSuggestApi;
	private OkHttpClient mClient;

	private Scheduler mObserveOn;
	private Scheduler mSubscribeOn;
	private static final int MAX_NEARBY_AIRPORTS = 2;

	private static final RequestInterceptor REQUEST_INTERCEPTOR = new RequestInterceptor() {
		@Override
		public void intercept(RequestFacade request) {
			request.addHeader("Accept", "application/json");
		}
	};

	public SuggestionServices(String endpoint, OkHttpClient okHttpClient, Scheduler observeOn, Scheduler subscribeOn, RestAdapter.LogLevel logLevel) {
		mObserveOn = observeOn;
		mSubscribeOn = subscribeOn;
		mClient = okHttpClient;

		Gson gson = new GsonBuilder()
			.registerTypeAdapter(SuggestionResponse.class, new SuggestionResponse())
			.create();

		RestAdapter adapter = new RestAdapter.Builder()
			.setEndpoint(endpoint)
			.setLogLevel(logLevel)
			.setConverter(new GsonConverter(gson))
			.setClient(new OkClient(mClient))
			.setRequestInterceptor(REQUEST_INTERCEPTOR)
			.build();

		mSuggestApi = adapter.create(SuggestApi.class);
	}

	private static final int MAX_AIRPORTS_RETURNED = 3;

	public Subscription getAirportSuggestions(String query, Observer<List<Suggestion>> observer) {
		return mSuggestApi.suggestV3(query, SuggestionResultType.AIRPORT, null)
			.observeOn(mObserveOn)
			.subscribeOn(mSubscribeOn)
			.concatMap(FLATTEN_SUGGESTIONS)
			.take(MAX_AIRPORTS_RETURNED)
			.toList()
			.subscribe(observer);
	}

	private static final int MAX_LX_SUGGESTIONS_RETURNED = 3;

	public Subscription getLxSuggestions(String query, Observer<List<Suggestion>> observer) {
		return mSuggestApi.suggestV3(query,
			SuggestionResultType.CITY | SuggestionResultType.MULTI_CITY | SuggestionResultType.NEIGHBORHOOD,
			"ACTIVITIES")
			.observeOn(mObserveOn)
			.subscribeOn(mSubscribeOn)
			.flatMap(FLATTEN_SUGGESTIONS)
			.take(MAX_LX_SUGGESTIONS_RETURNED)
			.toList()
			.subscribe(observer);
	}

	private static final Func1<SuggestionResponse, Observable<Suggestion>> FLATTEN_SUGGESTIONS = new Func1<SuggestionResponse, Observable<Suggestion>>() {
		@Override
		public Observable<Suggestion> call(SuggestionResponse suggestionResponse) {
			return Observable.from(suggestionResponse.suggestions);
		}
	};

	private static Func1<SuggestionResponse, List<Suggestion>> sToListNearby = new Func1<SuggestionResponse, List<Suggestion>>() {
		@Override
		public List<Suggestion> call(SuggestionResponse suggestionResponse) {
			if (suggestionResponse != null) {
				List<Suggestion> result = suggestionResponse.suggestions.subList(0, suggestionResponse.suggestions.size() >= 2 ? MAX_NEARBY_AIRPORTS : 1);
				for (Suggestion suggestion : result) {
					suggestion.iconType = Suggestion.IconType.CURRENT_LOCATION_ICON;
				}
				return result;
			}
			return null;
		}
	};

	public Subscription getNearbyAirportSuggestions(String locale, String latlong, int siteId, Observer<List<Suggestion>> observer) {
		return mSuggestApi.suggestNearbyV1(locale, latlong, siteId, SuggestionResultType.AIRPORT, "p")
			.observeOn(mObserveOn)
			.subscribeOn(mSubscribeOn)
			.map(sToListNearby)
			.subscribe(observer);
	}

	public Subscription getNearbyLxSuggestions(String locale, String latlong, int siteId, Observer<List<Suggestion>> observer) {
		return getNearbyLxSuggestions(locale, latlong, siteId)
			.subscribe(observer);
	}

	public Observable<List<Suggestion>> getNearbyLxSuggestions(String locale, String latlong, int siteId) {
		return mSuggestApi.suggestNearbyV1(locale, latlong, siteId,
			SuggestionResultType.CITY | SuggestionResultType.MULTI_CITY | SuggestionResultType.NEIGHBORHOOD, "d")
			.observeOn(mObserveOn)
			.subscribeOn(mSubscribeOn)
			.map(sToListNearby);
	}
}

File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/services/SuggestionServices.kt
package com.expedia.bookings.services

import com.expedia.bookings.data.SuggestionResultType
import com.expedia.bookings.data.SuggestionV4
import com.expedia.bookings.data.cars.SuggestionResponse
import com.google.gson.GsonBuilder
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import rx.Observable
import rx.Observer
import rx.Scheduler
import rx.Subscription
import java.util.Collections
import java.util.Comparator

class SuggestionServices(endpoint: String, okHttpClient: OkHttpClient, interceptor: Interceptor, val observeOn: Scheduler, val subscribeOn: Scheduler) {

    val suggestApi: SuggestApi by lazy {
        val gson = GsonBuilder().registerTypeAdapter(SuggestionResponse::class.java, SuggestionResponse()).create()

        val adapter = Retrofit.Builder()
                .baseUrl(endpoint)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .client(okHttpClient.newBuilder().addInterceptor(interceptor).build())
                .build()


        adapter.create<SuggestApi>(SuggestApi::class.java)
    }

    private val MAX_NEARBY_SUGGESTIONS = 3

    fun getCarSuggestions(query: String, locale: String, client: String, observer: Observer<MutableList<SuggestionV4>>): Subscription {
        val type = getTypeForCarSuggestions()
        val lob = "CARS";
        val features = "cars_rental"
        return suggestV4(query, type, locale, features, client, lob, null)
                .doOnNext { list -> sortCarSuggestions(list) }
                .subscribe(observer);
    }

    fun getNearbyCarSuggestions(locale: String, latlng: String, siteId: Int, client: String, observer: Observer<MutableList<SuggestionV4>>): Subscription {
        val type = getTypeForCarSuggestions()
        val sort = "p"
        var lob = "CARS"
        return suggestNearbyV4(locale, latlng, siteId, type, sort, client, lob)
                .doOnNext { list -> sortCarSuggestions(list) }
                .doOnNext { list -> renameFirstResultIdToCurrentLocation(list) }
                .subscribe(observer)
    }

    private fun renameFirstResultIdToCurrentLocation(suggestions: MutableList<SuggestionV4>): List<SuggestionV4> {
        if (suggestions.size > 0) {
            suggestions[0].gaiaId = SuggestionV4.CURRENT_LOCATION_ID
        }
        return suggestions
    }

    private fun getTypeForCarSuggestions(): Int {
        val type = SuggestionResultType.AIRPORT or SuggestionResultType.CITY or SuggestionResultType.MULTI_CITY or
                SuggestionResultType.NEIGHBORHOOD or SuggestionResultType.POINT_OF_INTEREST or SuggestionResultType.AIRPORT_METRO_CODE

        return type;
    }

    private fun sortCarSuggestions(suggestions: MutableList<SuggestionV4>) {
        Collections.sort(suggestions, object : Comparator<SuggestionV4> {
            override fun compare(lhs: SuggestionV4, rhs: SuggestionV4): Int {
                val leftSuggestionPrecedenceOrder = if (lhs.isMajorAirport) 1 else 2
                val rightSuggestionPrecedenceOrder = if (rhs.isMajorAirport) 1 else 2
                return leftSuggestionPrecedenceOrder.compareTo(rightSuggestionPrecedenceOrder)
            }
        })
    }

    fun getLxSuggestions(query: String, locale: String, client: String, observer: Observer<MutableList<SuggestionV4>>): Subscription {
        val type = SuggestionResultType.CITY or SuggestionResultType.MULTI_CITY or SuggestionResultType.NEIGHBORHOOD or
                SuggestionResultType.POINT_OF_INTEREST
        val lob = "ACTIVITIES"
        val features = "ta_hierarchy"
        return suggestV4(query, type, locale, features, client, lob, null).subscribe(observer)
    }

    private fun suggestV4(query: String, regiontype: Int, locale: String, features: String, client: String, lob: String?, siteId: Int?): Observable<MutableList<SuggestionV4>> {
        return suggestApi.suggestV4(query, locale, regiontype, false, features, client, lob, siteId)
                .observeOn(observeOn)
                .subscribeOn(subscribeOn)
                .map { response -> response.suggestions}
    }

    fun getNearbyLxSuggestions(locale: String, latlng: String, siteId: Int, client: String, observer: Observer<MutableList<SuggestionV4>>): Subscription {
        return getNearbyLxSuggestions(locale, latlng, siteId, client)
                .doOnNext { list -> renameFirstResultIdToCurrentLocation(list) }
                .subscribe(observer)
    }

    fun getNearbyLxSuggestions(locale: String, latlng: String, siteId: Int, client: String): Observable<MutableList<SuggestionV4>> {
        val type = SuggestionResultType.CITY or SuggestionResultType.MULTI_CITY or SuggestionResultType.NEIGHBORHOOD
        val sort = "d"
        val lob = "ACTIVITIES"
        return suggestNearbyV4(locale, latlng, siteId, type, sort, client, lob)
    }

    private fun suggestNearbyV4(locale: String, latlng: String, siteId: Int, type: Int, sort: String, client: String, lob: String): Observable<MutableList<SuggestionV4>> {
        return suggestApi.suggestNearbyV4(locale, latlng, siteId, type, sort, client, lob)
                .observeOn(observeOn)
                .subscribeOn(subscribeOn)
                .map { response -> response.suggestions.take(MAX_NEARBY_SUGGESTIONS) }
                .doOnNext { list ->
                    list.forEach {
                        suggestion ->
                        suggestion.iconType = SuggestionV4.IconType.CURRENT_LOCATION_ICON
                    }
                }
                .map {list -> list.toMutableList() }
    }
}
